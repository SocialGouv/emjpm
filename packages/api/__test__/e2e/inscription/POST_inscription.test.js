//

const request = require("supertest");

const nodemailerMock = require("nodemailer-mock");

jest.setMock("nodemailer", nodemailerMock);

// Fake env
process.env.SMTP_FROM = "ne-pas-repondre@emjpm.gouv.fr";
process.env.APP_URL = "https://emjpm.gouv.fr";

const { databaseName, knex } = global;
jest.setMock("@emjpm/api/src/db/knex", knex);

const server = require("@emjpm/api/src/server");
const { getAllTisByMandataire } = require("@emjpm/api/src/db/queries/tis");
const { seedData } = require("@emjpm/api/database/seed-data");

beforeAll(async () => {
  await seedData(databaseName);
});

beforeEach(async () => {
  nodemailerMock.mock.reset();
});

const defaultRegister = (params) => ({
  mandataire: {
    adresse: "",
    code_postal: "75010",
    department_id: 1,
    etablissement: "",
    telephone: "",
    ville: "",
  },
  user: {
    email: params && params.email != undefined ? params.email : "toto@toto.com",
    nom: params && params.nom != undefined ? params.nom : "testAd",
    password:
      params && params.password != undefined ? params.password : "test123456?",
    passwordConfirmation:
      params && params.passwordConfirmation != undefined
        ? params.passwordConfirmation
        : "test123456?",
    prenom: params && params.prenom != undefined ? params.prenom : "testPrenom",
    type: params && params.type != undefined ? params.type : "individuel",
    username: params && params.username != undefined ? params.username : "toto",
  },
});

test("should register with good values", async () => {
  const response = await request(server)
    .post("/api/auth/signup")
    .send({ ...defaultRegister(), tis: [22, 37] });
  expect(response.body).toMatchInlineSnapshot(`
                Object {
                  "success": true,
                }
        `);
  expect(response.status).toBe(200);

  const lastInsert = await knex("users").orderBy("created_at", "desc").first();
  const mandataire = await knex("mandataires")
    .orderBy("created_at", "desc")
    .first();
  const tis = await getAllTisByMandataire(mandataire.id);

  expect(tis.map((ti) => ti.id)).toEqual([22, 37]);
  expect(nodemailerMock.mock.sentMail().length).toBe(1);
  expect(nodemailerMock.mock.sentMail()).toMatchSnapshot();
  expect(lastInsert).toMatchSnapshot({
    created_at: expect.any(Object),
    password: expect.any(String),
    username: "toto",
  });
});

test("should NOT register when password!==passwordConfirmation", async () => {
  const response = await request(server)
    .post("/api/auth/signup")
    .send(
      defaultRegister({ password: "hello", passwordConfirmation: "world" })
    );

  expect(response.body).toMatchInlineSnapshot(
    { errors: expect.any(Array) },
    `
                    Object {
                      "errors": Any<Array>,
                    }
          `
  );
  expect(response.status).toBe(400);
  expect(nodemailerMock.mock.sentMail().length).toBe(0);
});

test("should NOT register when email already exist", async () => {
  const response = await request(server)
    .post("/api/auth/signup")
    .send(defaultRegister());

  expect(response.body).toMatchInlineSnapshot(
    { errors: expect.any(Array) },
    `
                        Object {
                          "errors": Any<Array>,
                        }
            `
  );
  expect(nodemailerMock.mock.sentMail().length).toBe(0);
  expect(response.status).toBe(409);
});

test("should add magistrat", async () => {
  const response = await request(server)
    .post("/api/auth/signup")
    .send({
      ...defaultRegister({
        email: "magistrat@emjpm.fr",
        type: "ti",
        username: "magistrat",
      }),
      magistrat: {
        cabinet: "2A",
        ti: 22,
      },
    });

  expect(response.body).toMatchInlineSnapshot(`
    Object {
      "success": true,
    }
  `);
  expect(response.status).toBe(200);
  expect(nodemailerMock.mock.sentMail()).toMatchSnapshot();

  const user = await knex.table("users").orderBy("created_at", "desc").first();

  expect(user.username).toEqual("magistrat");
  const magistrat = await knex
    .table("magistrat")
    .innerJoin("users", "users.id", "magistrat.user_id")
    .where("user_id", user.id)
    .first();
  expect(magistrat.cabinet).toEqual("2A");
  expect(magistrat.ti_id).toEqual(22);
});
