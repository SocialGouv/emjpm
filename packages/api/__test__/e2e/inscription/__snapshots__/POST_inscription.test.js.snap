// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add magistrat 2`] = `
Array [
  Object {
    "bcc": undefined,
    "from": "ne-pas-repondre@emjpm.gouv.fr",
    "html": "
Bonjour la super team eMJPM :),<br />
Nous avons reçu une nouvelle inscription pour un utilisateur:<br />
- type: ti <br />
- nom: testPrenom testAd <br />
- email: magistrat@emjpm.fr <br />
- code postal: 75010 <br /><br />

Merci de vérifier cette nouvelle demande et de la valider.
<br /><br />
Bonne journée :)
",
    "subject": "Nouvelle inscription",
    "text": "
Bonjour la super team emjpm :) ,

Nous avons reçu une nouvelle inscription pour un utilisateur:
- type: ti
- nom: testPrenom testAd
- email: magistrat@emjpm.fr
- code postal: 75010

Merci de vérifier cette nouvelle demande et de la valider.

Bonne journée :)
",
    "to": "support.emjpm@fabrique.social.gouv.fr",
  },
]
`;

exports[`should register with good values 2`] = `
Array [
  Object {
    "bcc": undefined,
    "from": "ne-pas-repondre@emjpm.gouv.fr",
    "html": "
Bonjour la super team eMJPM :),<br />
Nous avons reçu une nouvelle inscription pour un utilisateur:<br />
- type: individuel <br />
- nom: testPrenom testAd <br />
- email: toto@toto.com <br />
- code postal: 75010 <br /><br />

Merci de vérifier cette nouvelle demande et de la valider.
<br /><br />
Bonne journée :)
",
    "subject": "Nouvelle inscription",
    "text": "
Bonjour la super team emjpm :) ,

Nous avons reçu une nouvelle inscription pour un utilisateur:
- type: individuel
- nom: testPrenom testAd
- email: toto@toto.com
- code postal: 75010

Merci de vérifier cette nouvelle demande et de la valider.

Bonne journée :)
",
    "to": "support.emjpm@fabrique.social.gouv.fr",
  },
]
`;

exports[`should register with good values 3`] = `
Object {
  "active": false,
  "cabinet": null,
  "created_at": Any<Object>,
  "email": "toto@toto.com",
  "id": 2379,
  "last_login": null,
  "nom": "testAd",
  "password": Any<String>,
  "prenom": "testPrenom",
  "reset_password_expires": null,
  "reset_password_token": null,
  "type": "individuel",
}
`;
